

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Getting started with stanscofi &mdash; stanscofi v2.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=d2262ce8"></script>
      <script src="_static/doctools.js?v=888ff710"></script>
      <script src="_static/sphinx_highlight.js?v=4825356b"></script>
      <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Documentation in stanscofi" href="stanscofi_modules.html" />
    <link rel="prev" title="Installation of stanscofi" href="stanscofi_install.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            stanscofi
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="stanscofi_install.html">Installation of <em>stanscofi</em></a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Getting started with <em>stanscofi</em></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#handling-a-stanscofi-dataset">Handling a <em>stanscofi.Dataset</em></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#building-a-stanscofi-dataset-from-existing-datasets">Building a <em>stanscofi.Dataset</em> from existing datasets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#build-a-stanscofi-dataset-from-a-custom-dataset">Build a <em>stanscofi.Dataset</em> from a custom dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualization-with-a-stanscofi-dataset">Visualization with a <em>stanscofi.Dataset</em></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#using-ground-truth-annotations">Using ground truth annotations</a></li>
<li class="toctree-l4"><a class="reference internal" href="#using-predicted-annotations">Using predicted annotations</a></li>
<li class="toctree-l4"><a class="reference internal" href="#using-t-sne-or-umap-instead-of-pca">Using t-SNE or UMAP instead of PCA</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#create-weakly-correlated-training-and-testing-sets-from-a-stanscofi-dataset">Create weakly correlated training and testing sets from a <em>stanscofi.Dataset</em></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#train-a-stanscofi-model-example-of-nmf-on-binary-labelled-dataset">Train a <em>stanscofi.Model</em>: Example of NMF on binary labelled dataset</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#nmf-model">NMF model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fit-a-stanscofi-model">Fit a <em>stanscofi.Model</em></a></li>
<li class="toctree-l3"><a class="reference internal" href="#predict-from-a-stanscofi-model">Predict from a <em>stanscofi.Model</em></a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualize-the-predictions">Visualize the predictions</a></li>
<li class="toctree-l3"><a class="reference internal" href="#validate-a-fitted-stanscofi-model">Validate a fitted <em>stanscofi.Model</em></a></li>
<li class="toctree-l3"><a class="reference internal" href="#perform-cross-validation">Perform cross-validation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#perform-grid-search-on-hyperparameters">Perform grid-search on hyperparameters</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#extension-to-tri-labelled-datasets">Extension to tri-labelled datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preprocessing-approaches">Preprocessing approaches</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#concatenate-mean-impute-and-center-standardize-data">Concatenate, mean impute and center-standardize data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#train-a-stanscofi-model-example-of-logisticregression">Train a <em>stanscofi.Model</em>: Example of LogisticRegression</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#logistic-model">Logistic model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#model-fitting">Model fitting</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ranking-metrics">Ranking metrics</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#dcg-ncdg">DCG / NCDG</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mean-reciprocal-rank-mrr">Mean reciprocal rank (MRR)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#r-precision">R-Precision</a></li>
<li class="toctree-l3"><a class="reference internal" href="#precision-k">Precision &#64; k</a></li>
<li class="toctree-l3"><a class="reference internal" href="#average-precision-ap">Average Precision (AP)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mean-average-precision-map">Mean Average Precision (mAP)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#kendall-s-tau">Kendall’s <span class="math notranslate nohighlight">\(\tau\)</span></a></li>
<li class="toctree-l3"><a class="reference internal" href="#spearman-s-r">Spearman’s <span class="math notranslate nohighlight">\(r\)</span></a></li>
<li class="toctree-l3"><a class="reference internal" href="#expected-reciprocal-rank-err">Expected reciprocal rank (ERR)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="stanscofi_modules.html">Documentation in <em>stanscofi</em></a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="benchscofi_install.html">Installation of <em>benchscofi</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="benchscofi_content.html">Getting started with <em>benchscofi</em></a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">stanscofi</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Getting started with <em>stanscofi</em></li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/stanscofi_content.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <a class="reference internal image-reference" href="_images/header+EU_rescale.jpg"><img alt="RECeSS project" src="_images/header%2BEU_rescale.jpg" style="width: 700px;" /></a>
<section id="getting-started-with-stanscofi">
<h1>Getting started with <em>stanscofi</em><a class="headerlink" href="#getting-started-with-stanscofi" title="Permalink to this heading"></a></h1>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading"></a></h2>
<p>The following section is an introduction to the functionalities of <em>stanscofi</em>. It is also available as an interactive <a class="reference external" href="https://github.com/RECeSS-EU-Project/stanscofi/blob/master/docs/Introduction%20to%20stanscofi.ipynb">notebook</a>. The Python package <em>stanscofi</em> proposes standard procedures to train, and to validate drug repurposing classifiers, along with assessing and visualizing their performance. It comprises method-agnostic training and validation procedures on several drug repurposing datasets present in the literature.</p>
<p>The proper implementation of these steps is crucial in order to avoid data leakage, <em>i</em>.*e*., when the model is learnt over information that should be unavailable at prediction time. This can avoided by building training and validation sets which are weakly correlated with respect to the drug and disease feature vectors (function <code class="docutils literal notranslate"><span class="pre">weakly_correlated_split</span></code> in module <em>stanscofi.training_testing</em>). The main performance metric is the area under the curve (AUC), which estimates the diagnostic ability of a recommender system better than accuracy in imbalanced datasets, and the <span class="math notranslate nohighlight">\(F_\beta\)</span> score (function <code class="docutils literal notranslate"><span class="pre">compute_metrics</span></code> in module <em>stanscofi.validation</em>). Since version 2.0.0, other ranking measures are implemented.</p>
<p>Let’s create the folder containing the datasets:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir -p datasets/
</pre></div>
</div>
<p>and import base librairies and off we go:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">time</span><span class="w"> </span><span class="kn">import</span> <span class="n">time</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.sparse</span><span class="w"> </span><span class="kn">import</span> <span class="n">coo_array</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">stanscofi</span>
<span class="n">random_state</span> <span class="o">=</span> <span class="mi">124565</span>
<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">random_state</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Note 1:</strong> As a general rule, all functions are documented, so one can check out the inputs and outputs of a function func by typing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">func</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Note 2:</strong> To mesure your environmental impact when using this package (in terms of carbon emissions), please run the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ codecarbon init
</pre></div>
</div>
<p>to initialize the CodeCarbon config. For more information about using CodeCarbon, please refer to the <a class="reference external" href="https://github.com/mlco2/codecarbon">official repository</a>.</p>
</section>
<section id="handling-a-stanscofi-dataset">
<h2>Handling a <em>stanscofi.Dataset</em><a class="headerlink" href="#handling-a-stanscofi-dataset" title="Permalink to this heading"></a></h2>
<p>A dataset for collaborative filtering-based drug repurposing comprises three matrices:</p>
<ul class="simple">
<li><p>A matrix filled with 0, 1, or -1 values, which represents the known associations between drugs and diseases. Typically, 1 (resp., -1) stands for a positive (resp., negative) drug-disease matching. Finally, 0 represents unknown (untested) drug-disease matchings.</p></li>
<li><p>A matrix which encodes selected features for every drug involved in at least one matching in the previous matrix. Drug features might be the similarity of its chemical structure to the ones of other drugs, or the variation in gene activity which can be imputed to the treatment by that drug.</p></li>
<li><p>A matrix which similarly encodes features for every disease involved in at least one matching in the previous matrix. In that case, features of a disease might correspond to the similarity of its associated phenotypes (i.e., observable characteristics associated with the disease), or the variation in gene activity which can be imputed to the presence of disease.</p></li>
</ul>
<p>A dataset class with valuable methods (for the import and visualization) is implemented in <em>stanscofi.datasets</em>. The import of several drug repurposing datasets is implemented in <em>stanscofi.utils</em> (see <a class="reference external" href="https://github.com/RECeSS-EU-Project/drug-repurposing-datasets">this GitHub repository</a> for more information about the datasets).</p>
<p>This section shows how to use these methods.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">stanscofi.datasets</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">stanscofi.utils</span>
</pre></div>
</div>
<section id="building-a-stanscofi-dataset-from-existing-datasets">
<h3>Building a <em>stanscofi.Dataset</em> from existing datasets<a class="headerlink" href="#building-a-stanscofi-dataset-from-existing-datasets" title="Permalink to this heading"></a></h3>
<p>The first option to create a <em>stanscofi.Dataset</em> is to import a known drug repurposing dataset and convert it into a  <em>stanscofi.Dataset</em>. The available ones are shown in the following table:</p>
<table class="docutils align-default" id="id1">
<caption><span class="caption-text">Available datasets in <em>stanscofi</em></span><a class="headerlink" href="#id1" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 16.7%" />
<col style="width: 16.7%" />
<col style="width: 16.7%" />
<col style="width: 16.7%" />
<col style="width: 16.7%" />
<col style="width: 16.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Dataset</p></th>
<th class="head"><p>#drugs★ (#features)</p></th>
<th class="head"><p>#diseases★ (#features)</p></th>
<th class="head"><p>#positive matchings</p></th>
<th class="head"><p>#negative matchings</p></th>
<th class="head"><p>Sparsity number✦</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Gottlieb / FDataset</p></td>
<td><p>593 (593)</p></td>
<td><p>313 (313)</p></td>
<td><p>1,933</p></td>
<td><p>0</p></td>
<td><p>98.96%</p></td>
</tr>
<tr class="row-odd"><td><p>CDataset</p></td>
<td><p>663 (663)</p></td>
<td><p>409 (409)</p></td>
<td><p>2,532</p></td>
<td><p>0</p></td>
<td><p>99.07%</p></td>
</tr>
<tr class="row-even"><td><p>DNDataset</p></td>
<td><p>550 (1,490)</p></td>
<td><p>360 (4,516)</p></td>
<td><p>1,008</p></td>
<td><p>0</p></td>
<td><p>99.50%</p></td>
</tr>
<tr class="row-odd"><td><p>LRSSL</p></td>
<td><p>763 (2,049)</p></td>
<td><p>681 (681)</p></td>
<td><p>3,051</p></td>
<td><p>0</p></td>
<td><p>99.41%</p></td>
</tr>
<tr class="row-even"><td><p>PREDICT-Gottlieb</p></td>
<td><p>593 (1,779)</p></td>
<td><p>313 (313)</p></td>
<td><p>1,933</p></td>
<td><p>0</p></td>
<td><p>98.96%</p></td>
</tr>
<tr class="row-odd"><td><p>TRANSCRIPT</p></td>
<td><p>204 (12,096)</p></td>
<td><p>116 (12,096)</p></td>
<td><p>401</p></td>
<td><p>11</p></td>
<td><p>98.26%</p></td>
</tr>
<tr class="row-even"><td><p>PREDICT (private version)</p></td>
<td><p>1,351 (6,265)</p></td>
<td><p>1,066 (2,914)</p></td>
<td><p>5,624</p></td>
<td><p>152</p></td>
<td><p>99.60%</p></td>
</tr>
</tbody>
</table>
<p>★ involved in at least one known drug-disease matching
✦ percentage of unknown matchings in the drug-disease association matrix (note that it takes into account <em>all</em> diseases and drugs in the dataset, even those which are not involved with any known matching)</p>
<p>Note that for PREDICT, per default, the downloaded version is the partial dataset publicly available on Zenodo. If the notebook <em>PREDICT_dataset.ipynb</em> in <a class="reference external" href="https://github.com/RECeSS-EU-Project/drug-repurposing-datasets">this repository</a> was not run, then function <em>load_dataset</em> will return a dataset with the following characteristics:</p>
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">Public version of the PREDICT dataset</span><a class="headerlink" href="#id2" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 16.7%" />
<col style="width: 16.7%" />
<col style="width: 16.7%" />
<col style="width: 16.7%" />
<col style="width: 16.7%" />
<col style="width: 16.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Dataset</p></th>
<th class="head"><p>#drugs★ (#features)</p></th>
<th class="head"><p>#diseases★ (#features)</p></th>
<th class="head"><p>#positive matchings</p></th>
<th class="head"><p>#negative matchings</p></th>
<th class="head"><p>Sparsity number✦</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>PREDICT (public version)</p></td>
<td><p>1,014 (1,642)</p></td>
<td><p>941 (1,490)</p></td>
<td><p>4,627</p></td>
<td><p>132</p></td>
<td><p>99.50%</p></td>
</tr>
</tbody>
</table>
<p>Those datasets can be retrieved using the following names:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">available_datasets</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Gottlieb&quot;</span><span class="p">,</span> <span class="s2">&quot;Cdataset&quot;</span><span class="p">,</span> <span class="s2">&quot;DNdataset&quot;</span><span class="p">,</span> <span class="s2">&quot;LRSSL&quot;</span><span class="p">,</span>
    <span class="s2">&quot;PREDICT_Gottlieb&quot;</span><span class="p">,</span> <span class="s2">&quot;TRANSCRIPT&quot;</span><span class="p">,</span> <span class="s2">&quot;PREDICT&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>Printing the contents of the datasets should match the contents of the previous table.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">model_name</span> <span class="ow">in</span> <span class="n">available_datasets</span><span class="p">:</span>
   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;$ &quot;</span><span class="o">+</span><span class="n">model_name</span><span class="p">)</span>
   <span class="n">data_args</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">model_name</span><span class="p">,</span> <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">)</span>
   <span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">data_args</span><span class="p">)</span>
   <span class="c1"># Prints the summary of the dataset</span>
   <span class="n">dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="build-a-stanscofi-dataset-from-a-custom-dataset">
<h3>Build a <em>stanscofi.Dataset</em> from a custom dataset<a class="headerlink" href="#build-a-stanscofi-dataset-from-a-custom-dataset" title="Permalink to this heading"></a></h3>
<p>The second option to create a <em>stanscofi.Dataset</em> is to build one from scratch, using a matrix of ratings, a drug (“item”) feature matrix and a disease (“user”) feature matrix. Method <code class="docutils literal notranslate"><span class="pre">.summary</span></code> returns several measures of interest for the dataset (see <code class="docutils literal notranslate"><span class="pre">help</span></code>).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">di</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="s2">&quot;TRANSCRIPT&quot;</span><span class="p">,</span> <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">)</span>

<span class="n">ratings</span> <span class="o">=</span> <span class="n">di</span><span class="p">[</span><span class="s2">&quot;ratings&quot;</span><span class="p">]</span>
<span class="n">users</span> <span class="o">=</span> <span class="n">di</span><span class="p">[</span><span class="s2">&quot;users&quot;</span><span class="p">]</span>
<span class="n">items</span> <span class="o">=</span> <span class="n">di</span><span class="p">[</span><span class="s2">&quot;items&quot;</span><span class="p">]</span>

<span class="k">for</span> <span class="n">df_name</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ratings&quot;</span><span class="p">,</span> <span class="s2">&quot;users&quot;</span><span class="p">,</span> <span class="s2">&quot;items&quot;</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;$ &quot;</span><span class="o">+</span><span class="n">df_name</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">eval</span><span class="p">(</span><span class="n">df_name</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">5</span><span class="p">,:</span><span class="mi">5</span><span class="p">])</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">df_name</span><span class="o">==</span><span class="s2">&quot;ratings&quot;</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Values </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="nb">eval</span><span class="p">(</span><span class="n">df_name</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">)))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Shape (</span><span class="si">%d</span><span class="s2">, </span><span class="si">%d</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="nb">eval</span><span class="p">(</span><span class="n">df_name</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="n">ratings</span><span class="o">=</span><span class="n">ratings</span><span class="p">,</span> <span class="n">users</span><span class="o">=</span><span class="n">users</span><span class="p">,</span> <span class="n">items</span><span class="o">=</span><span class="n">items</span><span class="p">,</span>
       <span class="n">same_item_user_features</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;TRANSCRIPT &lt;custom loading&gt;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;$ &quot;</span><span class="o">+</span><span class="n">dataset</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="visualization-with-a-stanscofi-dataset">
<h3>Visualization with a <em>stanscofi.Dataset</em><a class="headerlink" href="#visualization-with-a-stanscofi-dataset" title="Permalink to this heading"></a></h3>
<p>Let us create a simple dataset (in the sense that positive and negative pairs are easily distinguishable from each other based on features). Each of the <span class="math notranslate nohighlight">\(N_f\)</span> features for (item, user) pair feature vectors associated with positive ratings are drawn from a Gaussian distribution of mean <span class="math notranslate nohighlight">\(m=0.5\)</span> and standard deviation <span class="math notranslate nohighlight">\(\sigma=1\)</span>, whereas those for negative ratings are drawn from from a Gaussian distribution of mean <span class="math notranslate nohighlight">\(m=0.5\)</span> and standard deviation <span class="math notranslate nohighlight">\(\sigma=1\)</span>.</p>
<p>We generate for “drugs” and “diseases” two matrices of shape (<span class="math notranslate nohighlight">\(\lfloor N_f/2 \rfloor, N_\text{pos}+N_\text{neg}\)</span>), which are the concatenation of <span class="math notranslate nohighlight">\(N_\text{pos}\)</span> positive and <span class="math notranslate nohighlight">\(N_\text{neg}\)</span> negative pair feature vectors generated from Gaussian distributions. Thus there are <span class="math notranslate nohighlight">\(N_\text{pos}^2\)</span> positive ratings (each “positive” disease with a “positive” drug), <span class="math notranslate nohighlight">\(N_\text{neg}^2\)</span> negative ratings (<em>idem</em>), and the remainder is unknown (that is, <span class="math notranslate nohighlight">\((N_\text{pos}+N_\text{neg})^2-N_\text{pos}^2-N_\text{neg}^2\)</span> ratings).</p>
<p>The function which generates this dataset is <em>stanscofi.datasets.generate_dummy_dataset</em>, and might be useful for testing purposes.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Parameters</span>
<span class="n">npositive</span><span class="p">,</span> <span class="n">nnegative</span><span class="p">,</span> <span class="n">nfeatures</span><span class="p">,</span> <span class="n">mean</span><span class="p">,</span> <span class="n">std</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span>
<span class="n">ddi</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">generate_dummy_dataset</span><span class="p">(</span><span class="n">npositive</span><span class="p">,</span> <span class="n">nnegative</span><span class="p">,</span> <span class="n">nfeatures</span><span class="p">,</span> <span class="n">mean</span><span class="p">,</span> <span class="n">std</span><span class="p">)</span>

<span class="c1">## Import it as a stanscofi.Dataset</span>
<span class="n">ddi</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;same_item_user_features&quot;</span><span class="p">:</span><span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;Dummy&quot;</span><span class="p">})</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">ddi</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
</pre></div>
</div>
<section id="using-ground-truth-annotations">
<h4>Using ground truth annotations<a class="headerlink" href="#using-ground-truth-annotations" title="Permalink to this heading"></a></h4>
<p>The method <em>visualize</em> in class <em>stanscofi.Dataset</em> plots a Principal Component Analysis (PCA) that allows to project the drug-disease pairs associated with known or unknown matchings based on drug and disease features onto the first two principal components (in terms of variance). The PCA is computed on the concatenation of (feature-wise) mean-imputed (in the presence of missing values), filtered (considering the Top-$N$ features in terms of cross-sample variance), and centered-standardized drug and disease feature matrices.</p>
<p>According to how we have built the dataset, one should expect that one square of the 2D plane contains negative pairs, another positive pairs, and unknown matching pairs are located outside of these two squares.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;* Plot the dataset with only nonzero ground truth annotations (PCA)&quot;</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;* Plot the dataset with only nonzero ground truth annotations (UMAP)&quot;</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">dimred_args</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;n_neighbors&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">})</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;* Plot the dataset with all ground truth annotations (PCA)&quot;</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_visualization.png"><img alt="Visualization of dataset with ground truth labels" src="_images/dataset_visualization.png" style="width: 600px;" /></a>
</section>
<section id="using-predicted-annotations">
<h4>Using predicted annotations<a class="headerlink" href="#using-predicted-annotations" title="Permalink to this heading"></a></h4>
<p>Now, since the true labels of datapoints are usually unknown, let us assume that we made predictions about those labels, and that we want to observe how good those predictions are. Here, we generate random predictions, that is, we flip a coin, which assigns a negative label with probability <span class="math notranslate nohighlight">\(\pi=\frac{1}{16}\)</span> (and a positive one otherwise).</p>
<p>The first plot shows the predicted labels (red for negative labels, green for positive ones) for each point, where the shape of the marker is related to the ground truth label for that point (<code class="docutils literal notranslate"><span class="pre">+</span></code> if positive, <code class="docutils literal notranslate"><span class="pre">v</span></code> if negative, <code class="docutils literal notranslate"><span class="pre">.</span></code> if unknown).</p>
<p>The second plot emphasizes on prediction errors (green for correct predictions, red for incorrect ones), whereas the marker is still related to the ground truth label. It is restricted to datapoints (drug, disease pairs) which have a known matching (i.e., a rating in {-1,1}).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Generate random class predictions</span>
<span class="n">pi</span><span class="o">=</span><span class="mi">1</span><span class="o">/</span><span class="mi">16</span>
<span class="n">shp</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">shape</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">coo_array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="n">pi</span><span class="p">,</span><span class="mi">1</span><span class="o">-</span><span class="n">pi</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">shp</span><span class="p">)))</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;* Plot the dataset with random predictions&quot;</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">predictions</span><span class="o">=</span><span class="n">predictions</span><span class="p">,</span> <span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;* Plot the dataset with random predictions and color by error&quot;</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">predictions</span><span class="o">=</span><span class="n">predictions</span><span class="p">,</span> <span class="n">show_errors</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_visualization_pred.png"><img alt="Visualization of dataset with predicted annotations" src="_images/dataset_visualization_pred.png" style="width: 600px;" /></a>
<p>For more information, <code class="docutils literal notranslate"><span class="pre">help(dataset.visualize)</span></code> might be of help.</p>
</section>
<section id="using-t-sne-or-umap-instead-of-pca">
<h4>Using t-SNE or UMAP instead of PCA<a class="headerlink" href="#using-t-sne-or-umap-instead-of-pca" title="Permalink to this heading"></a></h4>
<p>Since version 2.0.0 of <em>stanscofi</em>, the PCA plot can be replaced by a <a class="reference external" href="https://arxiv.org/abs/1802.03426">Uniform Manifold Approximation and Projection (UMAP)</a> plot, which might give a better overview of the similarity between drug-disease pairs. Let’s compare the plots for the datasets TRANSCRIPT and PREDICT. For the sake of completeness, we also compare those plots with the <a class="reference external" href="https://jmlr.org/papers/v9/vandermaaten08a.html">t-distributed Stochastic Neighbor Embedding (t-SNE)</a> approach (not natively implemented in <em>stanscofi</em>).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">stanscofi.training_testing</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.manifold</span><span class="w"> </span><span class="kn">import</span> <span class="n">TSNE</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">stanscofi.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">meanimputation_standardize</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.lines</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mlines</span>
</pre></div>
</div>
<p>Let’s plot those three representions for dataset TRANSCRIPT:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dataset_name</span><span class="o">=</span><span class="s2">&quot;TRANSCRIPT&quot;</span>
<span class="n">data_args</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span> <span class="n">dataset_folder</span><span class="p">)</span>
<span class="n">data_args</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;same_item_user_features&quot;</span><span class="p">:</span> <span class="p">(</span><span class="n">dataset_name</span><span class="o">==</span><span class="s2">&quot;TRANSCRIPT&quot;</span><span class="p">)})</span>
<span class="n">real_dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">data_args</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Plot PCA&quot;</span><span class="p">)</span>
<span class="n">real_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Plot UMAP&quot;</span><span class="p">)</span>
<span class="n">real_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">dimred_args</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;n_neighbors&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">})</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Plot (Barnes-Hut) t-SNE&quot;</span><span class="p">)</span>
<span class="n">nvalues</span> <span class="o">=</span> <span class="n">real_dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">subselect_size</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="nb">min</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">5e7</span><span class="p">)</span><span class="o">//</span><span class="n">nvalues</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">nvalues</span><span class="p">))</span>
<span class="n">subselect_size</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span>
     <span class="n">subselect_size</span><span class="p">,</span>
     <span class="nb">min</span><span class="p">(</span><span class="n">real_dataset</span><span class="o">.</span><span class="n">users</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">real_dataset</span><span class="o">.</span><span class="n">items</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="p">)</span>
<span class="c1">## Preprocessed (item, user) pair feature matrix and corresponding outcome vector</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">meanimputation_standardize</span><span class="p">(</span><span class="n">real_dataset</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="n">subselect_size</span><span class="p">,</span>
     <span class="n">inf</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">markers</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">((</span><span class="n">real_dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">row</span><span class="p">,</span> <span class="n">real_dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">col</span><span class="p">))</span>
<span class="n">markers</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">markers</span><span class="p">,</span> <span class="n">y</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">all_pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[{</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="s2">&quot;r.&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="s2">&quot;g.&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="s2">&quot;y.&quot;</span><span class="p">}[</span><span class="n">k</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">,</span><span class="n">k</span> <span class="ow">in</span> <span class="n">markers</span><span class="o">.</span><span class="n">tolist</span><span class="p">()])</span>
<span class="n">all_pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">markers</span><span class="p">,</span> <span class="n">all_pairs</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">all_pairs</span><span class="p">[:,:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_pairs</span><span class="p">[:,:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="c1">## Fit the t-SNE model</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">TSNE</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="mf">30.0</span><span class="p">,</span> <span class="n">early_exaggeration</span><span class="o">=</span><span class="mf">12.0</span><span class="p">,</span> <span class="n">learning_rate</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span>
    <span class="n">n_iter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">n_iter_without_progress</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">min_grad_norm</span><span class="o">=</span><span class="mf">1e-07</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span>
    <span class="n">metric_params</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1234</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;barnes_hut&#39;</span><span class="p">,</span>
    <span class="n">angle</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">dimred_X</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
<span class="c1">## Plot the t-SNE embeddings</span>
<span class="c1">## Put points in the front layer</span>
<span class="n">layer</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;g.&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;r.&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;y.&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}</span>
<span class="c1">## More visible points</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;g.&quot;</span><span class="p">:</span> <span class="mf">0.75</span><span class="p">,</span> <span class="s2">&quot;r.&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;y.&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">}</span>
<span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="k">for</span> <span class="n">mkr</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ravel</span><span class="p">(</span><span class="n">all_pairs</span><span class="p">[:,</span><span class="mi">3</span><span class="p">]))</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
    <span class="n">all_pairs_k</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">all_pairs</span><span class="p">[:,</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="n">mkr</span><span class="p">)[:,</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">mkr</span><span class="o">==</span><span class="s2">&quot;y.&quot;</span><span class="p">):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">dimred_X</span><span class="p">[</span><span class="n">all_pairs_k</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span> <span class="n">dimred_X</span><span class="p">[</span><span class="n">all_pairs_k</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span>
           <span class="n">zorder</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">dimred_X</span><span class="p">[</span><span class="n">all_pairs_k</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span> <span class="n">dimred_X</span><span class="p">[</span><span class="n">all_pairs_k</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="n">mkr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
           <span class="n">marker</span><span class="o">=</span><span class="n">mkr</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">zorder</span><span class="o">=</span><span class="n">layer</span><span class="p">[</span><span class="n">mkr</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">[</span><span class="n">mkr</span><span class="p">])</span>
<span class="n">handles</span> <span class="o">=</span> <span class="p">[</span><span class="n">mlines</span><span class="o">.</span><span class="n">Line2D</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">color</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;r&#39;</span><span class="p">:</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;g&#39;</span><span class="p">:</span><span class="s1">&#39;green&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">:</span><span class="s1">&#39;yellow&#39;</span><span class="p">}[</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span>
            <span class="n">label</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;r&#39;</span><span class="p">:</span><span class="s1">&#39;[-1]&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">:</span><span class="s1">&#39;[ 0]&#39;</span><span class="p">,</span><span class="s1">&#39;g&#39;</span><span class="p">:</span><span class="s2">&quot;[ 1]&quot;</span><span class="p">}[</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
            <span class="n">markersize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
    <span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">all_pairs</span><span class="p">[:,</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">str</span><span class="p">))</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="k">if</span> <span class="p">(</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;y&quot;</span><span class="p">)]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;t-SNE C2&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;t-SNE C1&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;on </span><span class="si">%d</span><span class="s2"> features&quot;</span> <span class="o">%</span> <span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="o">//</span><span class="mi">2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">handles</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">,</span>
     <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.6</span><span class="p">,</span><span class="mf">0.9</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/PCA_TRANSCRIPT.png"><img alt="Visualization of TRANSCRIPT dataset with PCA" src="_images/PCA_TRANSCRIPT.png" style="width: 195px;" /></a>
<a class="reference internal image-reference" href="_images/UMAP_TRANSCRIPT.png"><img alt="Visualization of TRANSCRIPT dataset with UMAP" src="_images/UMAP_TRANSCRIPT.png" style="width: 195px;" /></a>
<a class="reference internal image-reference" href="_images/tSNE_TRANSCRIPT.png"><img alt="Visualization of TRANSCRIPT dataset with t-SNE" src="_images/tSNE_TRANSCRIPT.png" style="width: 270px;" /></a>
<p>Let’s do the same with the dataset PREDICT with the PCA and t-SNE plots:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dataset_name</span><span class="o">=</span><span class="s2">&quot;PREDICT&quot;</span>
<span class="n">data_args</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span> <span class="n">dataset_folder</span><span class="p">)</span>
<span class="n">data_args</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;same_item_user_features&quot;</span><span class="p">:</span> <span class="p">(</span><span class="n">dataset_name</span><span class="o">==</span><span class="s2">&quot;TRANSCRIPT&quot;</span><span class="p">)})</span>
<span class="n">real_dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">data_args</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Plot PCA&quot;</span><span class="p">)</span>
<span class="n">real_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Plot (Barnes-Hut) t-SNE&quot;</span><span class="p">)</span>
<span class="n">nvalues</span> <span class="o">=</span> <span class="n">real_dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">subselect_size</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="nb">min</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">5e7</span><span class="p">)</span><span class="o">//</span><span class="n">nvalues</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">nvalues</span><span class="p">))</span>
<span class="n">subselect_size</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span>
    <span class="n">subselect_size</span><span class="p">,</span>
    <span class="nb">min</span><span class="p">(</span><span class="n">real_dataset</span><span class="o">.</span><span class="n">users</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">real_dataset</span><span class="o">.</span><span class="n">items</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="p">)</span>
<span class="c1">## Preprocessed (item, user) pair feature matrix and corresponding outcome vector</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">meanimputation_standardize</span><span class="p">(</span><span class="n">real_dataset</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="n">subselect_size</span><span class="p">,</span>
    <span class="n">inf</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">markers</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">((</span><span class="n">real_dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">row</span><span class="p">,</span> <span class="n">real_dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">col</span><span class="p">))</span>
<span class="n">markers</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">markers</span><span class="p">,</span> <span class="n">y</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">all_pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[{</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="s2">&quot;r.&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="s2">&quot;g.&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="s2">&quot;y.&quot;</span><span class="p">}[</span><span class="n">k</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">,</span><span class="n">k</span> <span class="ow">in</span> <span class="n">markers</span><span class="o">.</span><span class="n">tolist</span><span class="p">()])</span>
<span class="n">all_pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">markers</span><span class="p">,</span> <span class="n">all_pairs</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">all_pairs</span><span class="p">[:,:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_pairs</span><span class="p">[:,:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">TSNE</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="mf">30.0</span><span class="p">,</span> <span class="n">early_exaggeration</span><span class="o">=</span><span class="mf">12.0</span><span class="p">,</span>
    <span class="n">learning_rate</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">n_iter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">n_iter_without_progress</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">min_grad_norm</span><span class="o">=</span><span class="mf">1e-07</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">metric_params</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1234</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;barnes_hut&#39;</span><span class="p">,</span> <span class="n">angle</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">dimred_X</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
<span class="c1">## Put points in the front layer</span>
<span class="n">layer</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;g.&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;r.&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;y.&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}</span>
<span class="c1">## More visible points</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;g.&quot;</span><span class="p">:</span> <span class="mf">0.75</span><span class="p">,</span> <span class="s2">&quot;r.&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;y.&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">}</span>
<span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="k">for</span> <span class="n">mkr</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ravel</span><span class="p">(</span><span class="n">all_pairs</span><span class="p">[:,</span><span class="mi">3</span><span class="p">]))</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
    <span class="n">all_pairs_k</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">all_pairs</span><span class="p">[:,</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="n">mkr</span><span class="p">)[:,</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">mkr</span><span class="o">==</span><span class="s2">&quot;y.&quot;</span><span class="p">):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">dimred_X</span><span class="p">[</span><span class="n">all_pairs_k</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span> <span class="n">dimred_X</span><span class="p">[</span><span class="n">all_pairs_k</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span>
            <span class="n">zorder</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">dimred_X</span><span class="p">[</span><span class="n">all_pairs_k</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span> <span class="n">dimred_X</span><span class="p">[</span><span class="n">all_pairs_k</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="n">mkr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">marker</span><span class="o">=</span><span class="n">mkr</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">zorder</span><span class="o">=</span><span class="n">layer</span><span class="p">[</span><span class="n">mkr</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">[</span><span class="n">mkr</span><span class="p">])</span>
<span class="n">handles</span> <span class="o">=</span> <span class="p">[</span><span class="n">mlines</span><span class="o">.</span><span class="n">Line2D</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">color</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;r&#39;</span><span class="p">:</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;g&#39;</span><span class="p">:</span><span class="s1">&#39;green&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">:</span><span class="s1">&#39;yellow&#39;</span><span class="p">}[</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span>
            <span class="n">label</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;r&#39;</span><span class="p">:</span><span class="s1">&#39;[-1]&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">:</span><span class="s1">&#39;[ 0]&#39;</span><span class="p">,</span><span class="s1">&#39;g&#39;</span><span class="p">:</span><span class="s2">&quot;[ 1]&quot;</span><span class="p">}[</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
            <span class="n">markersize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
    <span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">all_pairs</span><span class="p">[:,</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">str</span><span class="p">))</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="k">if</span> <span class="p">(</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;y&quot;</span><span class="p">)]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;t-SNE C2&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;t-SNE C1&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;on </span><span class="si">%d</span><span class="s2"> features&quot;</span> <span class="o">%</span> <span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="o">//</span><span class="mi">2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">handles</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">,</span>
    <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.6</span><span class="p">,</span><span class="mf">0.9</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/PCA_PREDICT.png"><img alt="Visualization of PREDICT dataset with PCA" src="_images/PCA_PREDICT.png" style="width: 280px;" /></a>
<a class="reference internal image-reference" href="_images/tSNE_PREDICT.png"><img alt="Visualization of PREDICT dataset with t-SNE" src="_images/tSNE_PREDICT.png" style="width: 400px;" /></a>
</section>
</section>
<section id="create-weakly-correlated-training-and-testing-sets-from-a-stanscofi-dataset">
<h3>Create weakly correlated training and testing sets from a <em>stanscofi.Dataset</em><a class="headerlink" href="#create-weakly-correlated-training-and-testing-sets-from-a-stanscofi-dataset" title="Permalink to this heading"></a></h3>
<p>As previously mentioned at the start of this section, it is important to ensure that the training and testing datasets are weakly correlated. <em>stanscofi</em> proposes an computationally tractable approach which focuses on splitting items between training and testing sets (such that the set of items in both sets are disjoint <em>and</em> weakly correlated), based on hierarchical clustering. In a nutshell, it consists in splitting the set of items/drugs into two groups of clusters, such that the number of elements in the group assigned to testing is roughly a fixed proportion of the initial dataset (in terms of known ratings). A more refined splitting also allows to split the set of users/diseases.</p>
<p>Then one can create a subset from an existing dataset using those folds using method <em>subset</em> in the class <em>stanscofi.Dataset</em>.</p>
<p>For more information, refer to <code class="docutils literal notranslate"><span class="pre">help(stanscofi.training_testing.weakly_correlated_split)</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">stanscofi.training_testing</span>

<span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;Cdataset&quot;</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span>
      <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">))</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">dataset_name</span>

<span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;cityblock&quot;</span>
<span class="n">test_size</span> <span class="o">=</span> <span class="mf">0.25</span>

<span class="c1">## Generates weakly correlated training and testing datasets</span>
<span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">test_folds</span><span class="p">),</span> <span class="n">dists</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">weakly_correlated_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
   <span class="n">test_size</span><span class="p">,</span> <span class="n">early_stop</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Distances train/test: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">train/train: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">test/test: </span><span class="si">%f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">dists</span><span class="p">)</span>

<span class="n">train_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">test_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Test&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Training set&quot;</span><span class="p">)</span>
<span class="n">train_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Testing set&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>

<span class="n">train_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_train.png"><img alt="Visualization of training dataset with ground truth with weakly correlated split" src="_images/dataset_train.png" style="width: 280px;" /></a>
<a class="reference internal image-reference" href="_images/dataset_test.png"><img alt="Visualization of testing dataset with ground truth with weakly correlated split" src="_images/dataset_test.png" style="width: 400px;" /></a>
<p>Check that the percentage of known ratings in the training and testing sets is correct</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span>
   <span class="n">train_dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">*</span><span class="mi">100</span><span class="o">/</span><span class="n">dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span>
   <span class="c1">## should be close to/slightly less than 25%</span>
   <span class="n">test_dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">*</span><span class="mi">100</span><span class="o">/</span><span class="n">dataset</span><span class="o">.</span><span class="n">folds</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="p">)</span>
</pre></div>
</div>
<p>One can also decide to stick to randomly generated splits, as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Generates training and testing datasets at random</span>
<span class="p">(</span><span class="n">train_f_r</span><span class="p">,</span> <span class="n">test_f_r</span><span class="p">),</span> <span class="n">dists_r</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">random_simple_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
                      <span class="n">test_size</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
<span class="c1">## Compare to the ones in weakly correlated folds</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;(random) Distances train/test: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">train/train: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">test/test: </span><span class="si">%f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">dists_r</span><span class="p">)</span>

<span class="n">train_dataset_random</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">train_f_r</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train random&quot;</span><span class="p">)</span>
<span class="n">test_dataset_random</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">test_f_r</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Test random&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Training set (random)&quot;</span><span class="p">)</span>
<span class="n">train_dataset_random</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Testing set (random)&quot;</span><span class="p">)</span>
<span class="n">test_dataset_random</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>

<span class="n">train_dataset_random</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">test_dataset_random</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_random_train.png"><img alt="Visualization of training dataset with ground truth with random split" src="_images/dataset_random_train.png" style="width: 280px;" /></a>
<a class="reference internal image-reference" href="_images/dataset_random_test.png"><img alt="Visualization of testing dataset with ground truth with random split" src="_images/dataset_random_test.png" style="width: 400px;" /></a>
<p>One can also use <em>scikit-learn</em> cross-validation fold generators in a straightforward way:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.model_selection</span><span class="w"> </span><span class="kn">import</span> <span class="n">StratifiedKFold</span>
<span class="n">cv_generator</span> <span class="o">=</span> <span class="n">StratifiedKFold</span><span class="p">(</span><span class="n">n_splits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>

<span class="n">cv_folds</span><span class="p">,</span> <span class="n">dist_list</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">random_cv_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
         <span class="n">cv_generator</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">train_folds_random_cv</span><span class="p">,</span> <span class="n">test_folds_random_cv</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cv_folds</span><span class="p">):</span>
   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Fold #</span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

   <span class="c1">## Compare to the ones in weakly correlated folds</span>
   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;(random CV) Distances train/test: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">train/train: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">test/test: </span><span class="si">%f</span><span class="s2">&quot;</span><span class="o">%</span> <span class="n">dist_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

   <span class="n">train_dataset_random_cv</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">train_folds_random_cv</span><span class="p">,</span>
            <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train random CV&quot;</span><span class="p">)</span>
   <span class="n">test_dataset_random_cv</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">test_folds_random_cv</span><span class="p">,</span>
            <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Test random CV&quot;</span><span class="p">)</span>

   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Training set (random CV)&quot;</span><span class="p">)</span>
   <span class="n">train_dataset_random_cv</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Testing set (random CV)&quot;</span><span class="p">)</span>
   <span class="n">test_dataset_random_cv</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>

   <span class="n">train_dataset_random_cv</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
   <span class="n">test_dataset_random_cv</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="train-a-stanscofi-model-example-of-nmf-on-binary-labelled-dataset">
<h2>Train a <em>stanscofi.Model</em>: Example of NMF on binary labelled dataset<a class="headerlink" href="#train-a-stanscofi-model-example-of-nmf-on-binary-labelled-dataset" title="Permalink to this heading"></a></h2>
<p>A binary labelled dataset is a dataset in which ratings are either 1 (positive) or 0 (unknown), that is, there are no reported negative matchings. We first consider a method which only leverages the matrix of ratings, which is called Non-Negative Factorization (NMF).</p>
<section id="nmf-model">
<h3>NMF model<a class="headerlink" href="#nmf-model" title="Permalink to this heading"></a></h3>
<p>The goal is to identify two matrices <span class="math notranslate nohighlight">\(W \in \mathbb{R}^{r \times N_s}\)</span>, <span class="math notranslate nohighlight">\(H \in \mathbb{R}^{r \times N_p}\)</span> for a fixed value of <span class="math notranslate nohighlight">\(r \in \mathbb{N}^*\)</span> such that the unknown full matrix of ratings <span class="math notranslate nohighlight">\(R\)</span> satisfies</p>
<div class="math notranslate nohighlight">
\[R = W^\top H\;.\]</div>
<p>Considering the known matrix of item-user matchings <span class="math notranslate nohighlight">\(A \in \{-1,0,1\}^{N_s \times N_p}\)</span>, matrices <span class="math notranslate nohighlight">\(W\)</span> and <span class="math notranslate nohighlight">\(H\)</span> are infered by optimizing on the following problem</p>
<div class="math notranslate nohighlight">
\[W,H = \arg\min_{w,h} \frac{1}{2}\|A-w^\top h\|^2_F\;, \text{ where }\|\cdot\|_F \text{ is the Frobenius norm for matrices}\;.\]</div>
<p>Then, considering <span class="math notranslate nohighlight">\(R = W^\top H\)</span>, collaborative filtering can be performed by considering for any unlabeled value <span class="math notranslate nohighlight">\(A_{i,j} := \begin{cases} +1 &amp; \text{ if } R_{i,j}&gt;\theta\\-1 &amp; \text{otherwise}\end{cases}\)</span>. In particular, this method can only be tested on the dataset it was trained on.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">stanscofi.models</span>
</pre></div>
</div>
</section>
<section id="fit-a-stanscofi-model">
<h3>Fit a <em>stanscofi.Model</em><a class="headerlink" href="#fit-a-stanscofi-model" title="Permalink to this heading"></a></h3>
<p>First, we show how to fit the model on the whole training dataset.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Model parameters</span>
<span class="n">N_s</span><span class="p">,</span> <span class="n">N_p</span> <span class="o">=</span> <span class="n">train_dataset</span><span class="o">.</span><span class="n">ratings</span><span class="o">.</span><span class="n">shape</span>
<span class="n">grid_search</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;n_components&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">N_s</span><span class="p">,</span> <span class="n">N_p</span><span class="p">))}</span> <span class="c1"># value of r</span>
<span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;init&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;solver&quot;</span><span class="p">:</span><span class="s1">&#39;cd&#39;</span><span class="p">,</span> <span class="s2">&quot;beta_loss&quot;</span><span class="p">:</span><span class="s1">&#39;frobenius&#39;</span><span class="p">,</span> <span class="s2">&quot;tol&quot;</span><span class="p">:</span><span class="mf">0.0001</span><span class="p">,</span> <span class="s2">&quot;max_iter&quot;</span><span class="p">:</span><span class="mi">1500</span><span class="p">,</span>
  <span class="s2">&quot;random_state&quot;</span><span class="p">:</span><span class="n">random_state</span><span class="p">,</span> <span class="s2">&quot;alpha_W&quot;</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span> <span class="s2">&quot;alpha_H&quot;</span><span class="p">:</span><span class="s1">&#39;same&#39;</span><span class="p">,</span> <span class="s2">&quot;l1_ratio&quot;</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span> <span class="s2">&quot;verbose&quot;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span>
  <span class="s2">&quot;shuffle&quot;</span><span class="p">:</span><span class="kc">False</span><span class="p">}</span>

<span class="n">params</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;n_components&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">})</span>

<span class="c1">## Initialize the model</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">NMF</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>

<span class="c1">## Fit the model on the whole training dataset (with a fixed random seed)</span>
<span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;Cdataset&quot;</span>
<span class="n">train_dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span>
    <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">))</span>
<span class="n">train_dataset</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">dataset_name</span>
<span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">train_dataset</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="n">random_state</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="predict-from-a-stanscofi-model">
<h3>Predict from a <em>stanscofi.Model</em><a class="headerlink" href="#predict-from-a-stanscofi-model" title="Permalink to this heading"></a></h3>
<p>Second, we show how to output scores (confidence scores in labelling the considered datapoint as positive), how to print information about those scores. Then, one can call method <em>classify</em> in order to output classes instead of scores. Given a score <span class="math notranslate nohighlight">\(s\)</span> and a threshold <span class="math notranslate nohighlight">\(\theta\)</span> (fed to the model using argument <code class="docutils literal notranslate"><span class="pre">threshold</span></code>) the classification is performed as follows:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\begin{cases} 1 &amp; \text{ if }s \geq \theta\:,\\ -1 &amp; \text{ otherwise}\:.\end{cases}\end{split}\]</div>
<p>In code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">scores</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_proba</span><span class="p">(</span><span class="n">train_dataset</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">print_scores</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">print_classification</span><span class="p">(</span><span class="n">predictions</span><span class="p">)</span>

<span class="c1">## Return the score-wise top k pairs: returns a list of (item id, user id, score)</span>
<span class="n">lst_k</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">recommend_k_pairs</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="c1">## lst_thres==lst_k</span>
<span class="n">lst_thres</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">recommend_k_pairs</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="n">lst_k</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="s2">&quot;Top-k=</span><span class="si">%d</span><span class="s2"> with score </span><span class="si">%.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lst_k</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">score</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span><span class="n">_</span><span class="p">,</span><span class="n">score</span> <span class="ow">in</span> <span class="n">lst_thres</span><span class="p">]))</span>
</pre></div>
</div>
</section>
<section id="visualize-the-predictions">
<h3>Visualize the predictions<a class="headerlink" href="#visualize-the-predictions" title="Permalink to this heading"></a></h3>
<p>Now we visualize the resulting predictions on the PCA plot:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">train_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">predictions</span><span class="o">=</span><span class="n">predictions</span><span class="p">,</span> <span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_test_pred.png"><img alt="Visualization of testing dataset with predictions" src="_images/dataset_test_pred.png" style="width: 400px;" /></a>
<p>If we want to visualize the errors in predictions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">train_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">predictions</span><span class="o">=</span><span class="n">predictions</span><span class="p">,</span> <span class="n">show_errors</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_test_error.png"><img alt="Visualization of testing dataset with errors in predictions" src="_images/dataset_test_error.png" style="width: 400px;" /></a>
</section>
<section id="validate-a-fitted-stanscofi-model">
<h3>Validate a fitted <em>stanscofi.Model</em><a class="headerlink" href="#validate-a-fitted-stanscofi-model" title="Permalink to this heading"></a></h3>
<p>The package automatically computes and plots relevant validation measures (AUC, <span class="math notranslate nohighlight">\(F_\beta\)</span> score, precision-recall and receiver operating curves). The metrics are computed <em>per user / disease</em>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">stanscofi.validation</span>
</pre></div>
</div>
<p>The <em>area under the curve</em> (AUC) estimates the diagnostic ability of a recommender system better than accuracy in imbalanced datasets. The <span class="math notranslate nohighlight">\(F_\beta\)</span>-score quantifies the performance of the classifier in terms of precision and recall in recommendations. The parameter <span class="math notranslate nohighlight">\(\beta\)</span> controls how much importance is given to precision with respect to recall (in particular, <span class="math notranslate nohighlight">\(\beta=1\)</span> means equal importance). <span class="math notranslate nohighlight">\(k\)</span> is the rank at which to consider a truncated ranking metric. All the metrics implemented in <em>stanscofi</em> (from version 2.0.0) are listed in <code class="docutils literal notranslate"><span class="pre">metrics_list</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">metrics</span><span class="p">,</span> <span class="n">plot_args</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">compute_metrics</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">train_dataset</span><span class="p">,</span>
     <span class="n">metrics</span><span class="o">=</span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">metrics_list</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">metrics</span>
</pre></div>
</div>
<p>The first two plots are directly linked to the AUC and <span class="math notranslate nohighlight">\(F_\beta\)</span> scores. The third figure allows to observe the distribution of scores (independently of parameter <code class="docutils literal notranslate"><span class="pre">threshold</span></code> in method <code class="docutils literal notranslate"><span class="pre">.predict</span></code>. The last one allows to observe the proportion of each label with respect to the ground truth.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">plot_metrics</span><span class="p">(</span><span class="o">**</span><span class="n">plot_args</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span>
    <span class="n">model_name</span><span class="o">=</span><span class="n">model</span><span class="o">.</span><span class="n">name</span><span class="o">+</span><span class="s2">&quot; &quot;</span><span class="o">+</span><span class="n">dataset</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_test_validation.png"><img alt="Validation plots for the NMF model" src="_images/dataset_test_validation.png" style="width: 600px;" /></a>
</section>
<section id="perform-cross-validation">
<h3>Perform cross-validation<a class="headerlink" href="#perform-cross-validation" title="Permalink to this heading"></a></h3>
<p>Training is more effective when combined with cross-validation. <em>stanscofi</em> allows to run cross-validation steps (on different splits) in parallel. This function is close to function <code class="docutils literal notranslate"><span class="pre">sklearn.model_selection.cross_validate</span></code> in <em>scikit-learn</em>, and contrary to the latter, uses our custom “per user” metrics.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">njobs</span><span class="p">,</span> <span class="n">nsplits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span>

<span class="n">start</span><span class="o">=</span><span class="n">time</span><span class="p">()</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">cv_training</span><span class="p">(</span><span class="n">stanscofi</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">NMF</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">train_dataset</span><span class="p">,</span>
   <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;AUC&quot;</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="n">njobs</span><span class="p">,</span> <span class="n">nsplits</span><span class="o">=</span><span class="n">nsplits</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
   <span class="n">cv_type</span><span class="o">=</span><span class="s2">&quot;random&quot;</span><span class="p">,</span> <span class="n">show_plots</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Elapsed time: </span><span class="si">%.2f</span><span class="s2"> sec. (njobs=</span><span class="si">%d</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">,</span> <span class="n">njobs</span><span class="p">))</span>

<span class="nb">print</span><span class="p">([</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">results</span><span class="p">])</span>

<span class="s2">&quot;Best AUC on Test </span><span class="si">%f</span><span class="s2"> (Best on Train </span><span class="si">%f</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="p">(</span>
   <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">]),</span>
   <span class="n">results</span><span class="p">[</span><span class="s2">&quot;train_metric&quot;</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">])]</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="perform-grid-search-on-hyperparameters">
<h3>Perform grid-search on hyperparameters<a class="headerlink" href="#perform-grid-search-on-hyperparameters" title="Permalink to this heading"></a></h3>
<p>The routine <em>grid_search</em> allows us to look for the best value of <span class="math notranslate nohighlight">\(r\)</span> in NMF, by running a cross-validation learning phase for each value of <span class="math notranslate nohighlight">\(r\)</span>, and selecting the value of <span class="math notranslate nohighlight">\(r\)</span> which achieved the highest score across all cross-validation steps.</p>
<p>This function is close to function <code class="docutils literal notranslate"><span class="pre">sklearn.model_selection.GridSearchCV</span></code> in <em>scikit-learn</em>, and contrary to the latter, uses our custom “per user” metrics.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">N_s</span><span class="p">,</span> <span class="n">N_p</span> <span class="o">=</span> <span class="n">train_dataset</span><span class="o">.</span><span class="n">ratings</span><span class="o">.</span><span class="n">shape</span>
<span class="n">grid_search</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;n_components&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">N_s</span><span class="p">,</span> <span class="n">N_p</span><span class="p">)</span><span class="o">//</span><span class="mi">3</span><span class="p">,</span> <span class="mi">10</span><span class="p">)}</span>
<span class="n">grid_search</span>

<span class="n">best_params</span><span class="p">,</span><span class="n">best_estimator</span><span class="p">,</span><span class="n">best_metrics</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">grid_search</span><span class="p">(</span><span class="n">grid_search</span><span class="p">,</span>
       <span class="n">stanscofi</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">NMF</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">train_dataset</span><span class="p">,</span>  <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;AUC&quot;</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
       <span class="n">nsplits</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1234</span><span class="p">,</span> <span class="n">show_plots</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">best_estimator</span><span class="p">)</span> <span class="c1">## trained best model</span>
<span class="n">scores_test</span> <span class="o">=</span> <span class="n">best_estimator</span><span class="o">.</span><span class="n">predict_proba</span><span class="p">(</span><span class="n">test_dataset</span><span class="p">)</span>
<span class="n">predictions_test</span> <span class="o">=</span> <span class="n">best_estimator</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">scores_test</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;#1 = </span><span class="si">%d</span><span class="se">\t</span><span class="s2">#-1 = </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="nb">tuple</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">predictions_test</span><span class="o">==</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">]]))</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Best AUC on Test </span><span class="si">%f</span><span class="s2"> (Train </span><span class="si">%f</span><span class="s2">)&quot;</span>
 <span class="o">%</span> <span class="p">(</span><span class="n">best_metrics</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">],</span> <span class="n">best_metrics</span><span class="p">[</span><span class="s2">&quot;train_metric&quot;</span><span class="p">]))</span>
<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;params&quot;</span><span class="p">:</span> <span class="n">best_params</span><span class="p">})</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</section>
</section>
<section id="extension-to-tri-labelled-datasets">
<h2>Extension to tri-labelled datasets<a class="headerlink" href="#extension-to-tri-labelled-datasets" title="Permalink to this heading"></a></h2>
<p>A tri-labelled dataset is a dataset in which ratings are either 1 (positive), -1 (negative) or 0 (unknown). The functions and methods shown above actually also work in that case, as demonstrated on the dataset TRANSCRIPT. First, we generate the two training and testing datasets:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;TRANSCRIPT&quot;</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span>
            <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">))</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">dataset_name</span>

<span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;cosine&quot;</span>
<span class="n">test_size</span> <span class="o">=</span> <span class="mf">0.3</span>

<span class="c1">## Generates training and testing datasets</span>
<span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">test_folds</span><span class="p">),</span> <span class="n">dists</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">weakly_correlated_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
            <span class="n">test_size</span><span class="p">,</span> <span class="n">early_stop</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Distances train/test: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">train/train: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">test/test: </span><span class="si">%f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">dists</span><span class="p">)</span>

<span class="n">train_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">test_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Training set&quot;</span><span class="p">)</span>
<span class="n">train_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Testing set&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>

<span class="n">train_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_trilabelled.png"><img alt="Visualization of testing dataset with three types of labels" src="_images/dataset_trilabelled.png" style="width: 700px;" /></a>
<p>Then we apply a grid search on the number of components to choose while training:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Model parameters</span>
<span class="n">N_s</span><span class="p">,</span> <span class="n">N_p</span> <span class="o">=</span> <span class="n">train_dataset</span><span class="o">.</span><span class="n">ratings</span><span class="o">.</span><span class="n">shape</span>
<span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;init&quot;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;solver&quot;</span><span class="p">:</span><span class="s1">&#39;cd&#39;</span><span class="p">,</span> <span class="s2">&quot;beta_loss&quot;</span><span class="p">:</span><span class="s1">&#39;frobenius&#39;</span><span class="p">,</span> <span class="s2">&quot;tol&quot;</span><span class="p">:</span><span class="mf">0.0001</span><span class="p">,</span> <span class="s2">&quot;max_iter&quot;</span><span class="p">:</span><span class="mi">2000</span><span class="p">,</span>
         <span class="s2">&quot;random_state&quot;</span><span class="p">:</span><span class="n">random_state</span><span class="p">,</span> <span class="s2">&quot;alpha_W&quot;</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span> <span class="s2">&quot;alpha_H&quot;</span><span class="p">:</span><span class="s1">&#39;same&#39;</span><span class="p">,</span> <span class="s2">&quot;l1_ratio&quot;</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span>
         <span class="s2">&quot;verbose&quot;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;shuffle&quot;</span><span class="p">:</span><span class="kc">False</span><span class="p">}</span>
<span class="n">params</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;n_components&quot;</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">N_s</span><span class="p">,</span><span class="n">N_p</span><span class="p">)</span><span class="o">//</span><span class="mi">3</span><span class="p">})</span>

<span class="c1">## Fit the model on the whole training dataset</span>
<span class="n">nsplits</span><span class="o">=</span><span class="mi">5</span>
<span class="n">njobs</span><span class="o">=</span><span class="n">nsplits</span><span class="o">-</span><span class="mi">1</span>

<span class="c1">## Choose between training on the whole dataset,</span>
<span class="c1">## applying cross-validation with default parameters</span>
<span class="c1">## or running a grid search on some of the parameters</span>
<span class="n">training</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;whole&quot;</span><span class="p">,</span> <span class="s2">&quot;cv&quot;</span><span class="p">,</span> <span class="s2">&quot;gridsearch&quot;</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>

<span class="c1">## Whole training dataset</span>
<span class="k">if</span> <span class="p">(</span><span class="n">training</span><span class="o">==</span><span class="s2">&quot;whole&quot;</span><span class="p">):</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">NMF</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>
    <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">train_dataset</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">params</span>
    <span class="n">cv</span> <span class="o">=</span> <span class="kc">None</span>
<span class="k">else</span><span class="p">:</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">training</span><span class="o">==</span><span class="s2">&quot;cv&quot;</span><span class="p">):</span>
        <span class="c1">## Cross-validation</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">cv_training</span><span class="p">(</span><span class="n">stanscofi</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">NMF</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span>
            <span class="n">train_dataset</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;AUC&quot;</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="n">njobs</span><span class="p">,</span> <span class="n">nsplits</span><span class="o">=</span><span class="n">nsplits</span><span class="p">,</span>
            <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span> <span class="n">cv_type</span><span class="o">=</span><span class="s2">&quot;random&quot;</span><span class="p">,</span> <span class="n">show_plots</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s2">&quot;models&quot;</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">])]</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">params</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;AUC on Test </span><span class="si">%f</span><span class="s2"> (Train </span><span class="si">%f</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span>
                <span class="n">results</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">]),</span>
                <span class="n">results</span><span class="p">[</span><span class="s2">&quot;train_metric&quot;</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">]</span>
              <span class="p">)]</span>
        <span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1">## Grid-search</span>
        <span class="n">grid_search</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;n_components&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">10</span><span class="p">]}</span>
        <span class="n">best_params</span><span class="p">,</span><span class="n">best_estimator</span><span class="p">,</span><span class="n">best_metrics</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">grid_search</span><span class="p">(</span>
            <span class="n">grid_search</span><span class="p">,</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">NMF</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">train_dataset</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;AUC&quot;</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="n">njobs</span><span class="p">,</span> <span class="n">nsplits</span><span class="o">=</span><span class="n">nsplits</span><span class="p">,</span>
            <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span> <span class="n">show_plots</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">best_estimator</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">best_params</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;AUC on Test </span><span class="si">%f</span><span class="s2"> (Train </span><span class="si">%f</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="p">(</span>
          <span class="n">best_metrics</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">],</span>
          <span class="n">best_metrics</span><span class="p">[</span><span class="s2">&quot;train_metric&quot;</span><span class="p">]</span>
        <span class="p">))</span>

<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;params&quot;</span><span class="p">:</span> <span class="n">args</span><span class="p">})</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
<p>In order to compute metrics on datasets with more than two classes, the argument <code class="docutils literal notranslate"><span class="pre">average='weighted'</span></code> is fed to the AUC and F-score functions in <em>scikit-learn</em>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Validate</span>
<span class="n">scores</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_proba</span><span class="p">(</span><span class="n">test_dataset</span><span class="p">)</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">metrics</span><span class="p">,</span> <span class="n">plot_args</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">compute_metrics</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">test_dataset</span><span class="p">,</span>
                               <span class="n">metrics</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;AUC&quot;</span><span class="p">,</span> <span class="s2">&quot;Fscore&quot;</span><span class="p">),</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">plot_metrics</span><span class="p">(</span><span class="o">**</span><span class="n">plot_args</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span>
    <span class="n">model_name</span><span class="o">=</span><span class="n">model</span><span class="o">.</span><span class="n">name</span><span class="o">+</span><span class="s2">&quot; &quot;</span><span class="o">+</span><span class="n">dataset</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="n">metrics</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_trilabelled_validation.png"><img alt="Validation on a dataset with three types of labels" src="_images/dataset_trilabelled_validation.png" style="width: 700px;" /></a>
<p>Then we visualize errors in predictions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Plot</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">predictions</span><span class="o">=</span><span class="n">predictions</span><span class="p">,</span> <span class="n">show_errors</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">dimred_args</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;n_neighbors&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">})</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_trilabelled_visualization.png"><img alt="Visualization of errors on a dataset with three types of labels" src="_images/dataset_trilabelled_visualization.png" style="width: 700px;" /></a>
</section>
<section id="preprocessing-approaches">
<h2>Preprocessing approaches<a class="headerlink" href="#preprocessing-approaches" title="Permalink to this heading"></a></h2>
<p>These functions allow to clean (missing or non finite values, etc.) and to prepare the drug repurposing dataset into a form that can be accepted by most models, that is, (<span class="math notranslate nohighlight">\(X\)</span>, <span class="math notranslate nohighlight">\(y\)</span>), where <span class="math notranslate nohighlight">\(X\)</span> is of shape (<code class="docutils literal notranslate"><span class="pre">n_ratings</span></code>, <code class="docutils literal notranslate"><span class="pre">n_features</span></code>) and <span class="math notranslate nohighlight">\(y\)</span> of shape (<code class="docutils literal notranslate"><span class="pre">n_ratings</span></code>, ).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">stanscofi.preprocessing</span>
</pre></div>
</div>
<section id="concatenate-mean-impute-and-center-standardize-data">
<h3>Concatenate, mean impute and center-standardize data<a class="headerlink" href="#concatenate-mean-impute-and-center-standardize-data" title="Permalink to this heading"></a></h3>
<p>It is actually the preprocessing approach used in order to plot the PCA in the class <em>stanscofi.Dataset</em>. It is the most basic approach when no assumption is made on the data.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data_di</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="s2">&quot;Gottlieb&quot;</span><span class="p">,</span> <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">)</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">data_di</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_preproc_cs.png"><img alt="Visualization of a dataset" src="_images/dataset_preproc_cs.png" style="width: 700px;" /></a>
<p>Let’s show that we obtain the same plot as the <code class="docutils literal notranslate"><span class="pre">.visualize</span></code> function by using this preprocessing step:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">start</span><span class="o">=</span><span class="n">time</span><span class="p">()</span>

<span class="c1">## With the same parameters as in stanscofi.Dataset.visualize</span>
<span class="n">nvalues</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">ratings</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="c1"># Top-N features in terms of variance will be selected</span>
<span class="n">subselect_size</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="nb">min</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">5e7</span><span class="p">)</span><span class="o">//</span><span class="n">nvalues</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">nvalues</span><span class="p">))</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">meanimputation_standardize</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
        <span class="n">subset</span><span class="o">=</span><span class="n">subselect_size</span><span class="p">,</span> <span class="n">inf</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Elapsed time </span><span class="si">%.2f</span><span class="s2"> sec&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">))</span>
<span class="n">X</span><span class="o">.</span><span class="n">shape</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_preproc_cs.png"><img alt="Visualization of a dataset preprocessed with the centered mean imputed method" src="_images/dataset_preproc_cs.png" style="width: 700px;" /></a>
<dl class="simple">
<dt>Transcriptomic data</dt><dd></dd>
</dl>
<p>However, for transcriptomic data (for instance, in TRANSCRIPT), the features are the same both in the drug and disease feature matrices. Indeed, features are genes here. We might be able to better exploit this by considering a drug-disease feature vector which is the list of coefficient-wise products of the corresponding drug and disease feature vectors. A product which is positive (resp., negative) means that the disease and the drug might have similar (resp., opposite) effects (and in the same direction) on the transcriptome, which is intuitively bad (resp., good) for a drug candidate for this disease.</p>
<p>Function <code class="docutils literal notranslate"><span class="pre">same_feature_preprocessing</span></code> can take as input any arithmetic operator such as <code class="docutils literal notranslate"><span class="pre">+</span></code>, <code class="docutils literal notranslate"><span class="pre">*</span></code>, <code class="docutils literal notranslate"><span class="pre">-</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;TRANSCRIPT&quot;</span>
<span class="n">data_di</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span> <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">)</span>
<span class="n">data_di</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">dataset_name</span><span class="p">,</span> <span class="s2">&quot;same_item_user_features&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">})</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">data_di</span><span class="p">)</span>

<span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;cosine&quot;</span>
<span class="n">test_size</span> <span class="o">=</span> <span class="mf">0.2</span>

<span class="c1">## Generates training and testing datasets</span>
<span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">test_folds</span><span class="p">),</span> <span class="n">dists</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">weakly_correlated_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
            <span class="n">test_size</span><span class="p">,</span> <span class="n">early_stop</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Distances train/test: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">train/train: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">test/test: </span><span class="si">%f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">dists</span><span class="p">)</span>

<span class="n">train_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">test_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Training set&quot;</span><span class="p">)</span>
<span class="n">train_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Testing set&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>

<span class="n">start</span><span class="o">=</span><span class="n">time</span><span class="p">()</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">same_feature_preprocessing</span><span class="p">(</span><span class="n">test_dataset</span><span class="p">,</span> <span class="s2">&quot;*&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Elapsed time </span><span class="si">%.2f</span><span class="s2"> sec&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">))</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">)</span>

<span class="n">start</span><span class="o">=</span><span class="n">time</span><span class="p">()</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">same_feature_preprocessing</span><span class="p">(</span><span class="n">test_dataset</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Elapsed time </span><span class="si">%.2f</span><span class="s2"> sec&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">))</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_preproc_times.png"><img alt="Visualization of a dataset preprocessed with the &quot;multiplicative&quot; approach" src="_images/dataset_preproc_times.png" style="width: 280px;" /></a>
<a class="reference internal image-reference" href="_images/dataset_preproc_add.png"><img alt="Visualization of a dataset preprocessed with the &quot;additive&quot; approach" src="_images/dataset_preproc_add.png" style="width: 400px;" /></a>
<dl class="simple">
<dt><a class="reference external" href="https://doi.org/10.1089/cmb.2010.0213">Perlman procedure</a></dt><dd></dd>
</dl>
<p>The paper cited in the title has proposed an approach to combining drug-drug and disease-disease similarity matrices so that the final feature vectors are much shorter than the ones obtained by simply concatenating those matrices. This approach was shown to improve the performance of recommender systems, and was used in combination with logistic regression in the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3159979/">PREDICT paper</a>.</p>
<p>In particular, it can be used for the following datasets: “Gottlieb”, “Cdataset”, “DNdataset”, “PREDICT_Gottlieb”, “PREDICT”.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;Gottlieb&quot;</span>
<span class="n">data_di</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span> <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">)</span>
<span class="n">data_di</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">dataset_name</span><span class="p">})</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">data_di</span><span class="p">)</span>

<span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;cityblock&quot;</span>
<span class="n">test_size</span> <span class="o">=</span> <span class="mf">0.1</span>

<span class="c1">## Generates training and testing datasets</span>
<span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">test_folds</span><span class="p">),</span> <span class="n">dists</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">weakly_correlated_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
             <span class="n">test_size</span><span class="p">,</span> <span class="n">early_stop</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Distances train/test: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">train/train: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">test/test: </span><span class="si">%f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">dists</span><span class="p">)</span>

<span class="n">train_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">test_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Training set&quot;</span><span class="p">)</span>
<span class="n">train_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Testing set&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>

<span class="n">start</span><span class="o">=</span><span class="n">time</span><span class="p">()</span>
<span class="c1">## set sep_feature to the separator between the feature type and the feature name</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">Perlman_procedure</span><span class="p">(</span><span class="n">test_dataset</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
              <span class="n">sep_feature</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Elapsed time </span><span class="si">%.2f</span><span class="s2"> sec&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">))</span>
<span class="n">X</span><span class="o">.</span><span class="n">shape</span>

<span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span>
              <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_perlman.png"><img alt="Visualization of a dataset preprocessed with the Perlman procedure" src="_images/dataset_perlman.png" style="width: 700px;" /></a>
<p>This approach can also be used for “LRSSL”, “TRANSCRIPT” if a similarity matrix is created from the feature matrices.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;LRSSL&quot;</span>
<span class="n">data_di</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span> <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">)</span>
<span class="n">data_di</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">dataset_name</span><span class="p">})</span>
<span class="c1">## Create correlation matrices</span>
<span class="n">data_di</span><span class="p">[</span><span class="s2">&quot;users&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_di</span><span class="p">[</span><span class="s2">&quot;users&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;pearson&quot;</span><span class="p">)</span>
<span class="n">data_di</span><span class="p">[</span><span class="s2">&quot;items&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_di</span><span class="p">[</span><span class="s2">&quot;items&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;pearson&quot;</span><span class="p">)</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">data_di</span><span class="p">)</span>

<span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;cityblock&quot;</span>
<span class="n">test_size</span> <span class="o">=</span> <span class="mf">0.3</span>

<span class="c1">## Generates training and testing datasets using a random split</span>
<span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">test_folds</span><span class="p">),</span> <span class="n">dists</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">random_simple_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
       <span class="n">test_size</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">12435</span><span class="p">)</span>

<span class="n">train_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">test_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Training set&quot;</span><span class="p">)</span>
<span class="n">train_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Testing set&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>

<span class="c1">## Preprocessing</span>
<span class="c1">## considering that there is a single similarity type ({1}), sep_feature=None (default value)</span>
<span class="n">start</span><span class="o">=</span><span class="n">time</span><span class="p">()</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">Perlman_procedure</span><span class="p">(</span><span class="n">test_dataset</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
       <span class="n">sep_feature</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Elapsed time </span><span class="si">%.2f</span><span class="s2"> sec (X shape </span><span class="si">%d</span><span class="s2">, </span><span class="si">%d</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">,</span> <span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

<span class="c1">## Result</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
<span class="c1">## Plots per dimension</span>
<span class="k">if</span> <span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">!=</span><span class="mi">1</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;* Dimension </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_LRSSL_perlman.png"><img alt="Visualization of the LRSSL dataset preprocessed with the Perlman procedure" src="_images/dataset_LRSSL_perlman.png" style="width: 700px;" /></a>
<p>Let’s try with the TRANSCRIPT dataset:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;TRANSCRIPT&quot;</span>
<span class="n">data_di</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span> <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">)</span>
<span class="n">data_di</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">dataset_name</span><span class="p">})</span>
<span class="c1">## Create correlation matrices</span>
<span class="n">data_di</span><span class="p">[</span><span class="s2">&quot;users&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_di</span><span class="p">[</span><span class="s2">&quot;users&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;pearson&quot;</span><span class="p">)</span>
<span class="n">data_di</span><span class="p">[</span><span class="s2">&quot;items&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_di</span><span class="p">[</span><span class="s2">&quot;items&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;pearson&quot;</span><span class="p">)</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">data_di</span><span class="p">)</span>

<span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;cosine&quot;</span>
<span class="n">test_size</span> <span class="o">=</span> <span class="mf">0.3</span>

<span class="c1">## Generates training and testing datasets at random</span>
<span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">test_folds</span><span class="p">),</span> <span class="n">dists</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">random_simple_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
    <span class="n">test_size</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">12435</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Distances train/test: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">train/train: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">test/test: </span><span class="si">%f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">dists</span><span class="p">)</span>

<span class="n">train_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">test_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Training set&quot;</span><span class="p">)</span>
<span class="n">train_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Testing set&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>

<span class="c1">## Preprocessing (considering that there is a single similarity type ({1}))</span>
<span class="n">start</span><span class="o">=</span><span class="n">time</span><span class="p">()</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">Perlman_procedure</span><span class="p">(</span><span class="n">test_dataset</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Elapsed time </span><span class="si">%.2f</span><span class="s2"> sec (X shape </span><span class="si">%d</span><span class="s2">, </span><span class="si">%d</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">,</span> <span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

<span class="c1">## Result</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
<span class="c1">## Plots per dimension</span>
<span class="k">if</span> <span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">!=</span><span class="mi">1</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;* Dimension </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">withzeros</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/dataset_TRANSCRIPT_perlman.png"><img alt="Visualization of the TRANSCRIPT dataset preprocessed with the Perlman procedure" src="_images/dataset_TRANSCRIPT_perlman.png" style="width: 700px;" /></a>
<p>There is a function <code class="docutils literal notranslate"><span class="pre">stanscofi.preprocessing.processing_XY</span></code> which can call those preprocessing functions in a straightforward way.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">stanscofi</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">processing_XY</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="train-a-stanscofi-model-example-of-logisticregression">
<h2>Train a <em>stanscofi.Model</em>: Example of LogisticRegression<a class="headerlink" href="#train-a-stanscofi-model-example-of-logisticregression" title="Permalink to this heading"></a></h2>
<p>Contrary to NMF, this model leverages the information from feature vectors.</p>
<section id="logistic-model">
<h3>Logistic model<a class="headerlink" href="#logistic-model" title="Permalink to this heading"></a></h3>
<p>If the considered model is a logistic model, then if the feature vector associated with drug-disease pair (i,j) is <span class="math notranslate nohighlight">\(x=[x_1, x_2, \dots, x_{L-1}, x_{L}]^\top\)</span> of length <span class="math notranslate nohighlight">\(L\)</span>, let us denote the set of <span class="math notranslate nohighlight">\(L+1\)</span> coefficients associated with positive class <span class="math notranslate nohighlight">\(C^+\)</span> (resp., negative class <span class="math notranslate nohighlight">\(C^-\)</span>) in the class prediction model <span class="math notranslate nohighlight">\(\beta^{C^+}_0\)</span>, <span class="math notranslate nohighlight">\(\beta^{C^+}_1\)</span>, <span class="math notranslate nohighlight">\(\dots\)</span>, <span class="math notranslate nohighlight">\(\beta^{C^+}_{L-1}\)</span>, <span class="math notranslate nohighlight">\(\beta^{C^+}_{L}\)</span> (resp., <span class="math notranslate nohighlight">\(\beta^{C^-}_0\)</span>, <span class="math notranslate nohighlight">\(\beta^{C^-}_1\)</span>, <span class="math notranslate nohighlight">\(\dots\)</span>, <span class="math notranslate nohighlight">\(\beta^{C^-}_{L-1}\)</span>, <span class="math notranslate nohighlight">\(\beta^{C^-}_{L}\)</span>). Then the probability of the associated cell of belonging to class <span class="math notranslate nohighlight">\(C\)</span> is</p>
<div class="math notranslate nohighlight">
\[Pr(x \text{ in } C | \beta^C_{\cdot}) := \sigma(\beta^C_0+\beta^C_1 x_1+\dots+\beta^C_{L-1} x_{L-1}+\beta^C_L x_L)\;, \text{ where } \sigma(x) := (1+\exp(-x))^{-1}\:.\]</div>
</section>
<section id="model-fitting">
<h3>Model fitting<a class="headerlink" href="#model-fitting" title="Permalink to this heading"></a></h3>
<p>Given <span class="math notranslate nohighlight">\(N\)</span> drug-disease pair feature vectors and associated ratings <span class="math notranslate nohighlight">\((x^i, r^i), i=1,2,\dots\)</span> (training data), the model fitting step consists to estimating the values of the <span class="math notranslate nohighlight">\(L\)</span> coefficients <span class="math notranslate nohighlight">\(\beta^C_0\)</span>, <span class="math notranslate nohighlight">\(\beta^C_1\)</span>, <span class="math notranslate nohighlight">\(\dots\)</span>, <span class="math notranslate nohighlight">\(\beta^C_{L-1}\)</span>, <span class="math notranslate nohighlight">\(\beta^C_{L}\)</span> for each class <span class="math notranslate nohighlight">\(C\)</span>. We consider a <span class="math notranslate nohighlight">\(\ell_1\)</span> (also called lasso) regularization, that is, the estimated set of parameters <span class="math notranslate nohighlight">\(\beta^{\cdot}_{\cdot}\)</span> that, for all classes and genes, would minimize the following function (<span class="math notranslate nohighlight">\(2L\)</span> coefficients are estimated)</p>
<div class="math notranslate nohighlight">
\[\begin{split}\text{loss}(\beta^{\cdot}_{\cdot}) := - \sum_{\text{class } C} \ \sum_{\substack{\text{pair }x^i,\\\text{ s.t. }r^i=C}} \  \log(Pr(x^i \text{ in } C | \beta^C_{\cdot}))+\sum_{\text{class } C} \ \sum_{\substack{\text{feature }\\f=1,2,\dots,L}} |\beta^{C}_{f}|\:.\end{split}\]</div>
<p>To perform model fitting, we use the <em>Logistic_Regression</em> class in <em>scikit-learn</em>. It is a bit slow when the number of drug-disease pairs and features is too large. As such, it should only be used for testing purposes, and/or in combination with the preprocessing procedures shown above.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;TRANSCRIPT&quot;</span>
<span class="n">data_di</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span> <span class="n">save_folder</span><span class="o">=</span><span class="s2">&quot;../datasets/&quot;</span><span class="p">)</span>
<span class="n">data_di</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">dataset_name</span><span class="p">,</span> <span class="s2">&quot;same_item_user_features&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">})</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="o">**</span><span class="n">data_di</span><span class="p">)</span>

<span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;cosine&quot;</span>
<span class="n">test_size</span> <span class="o">=</span> <span class="mf">0.1</span>

<span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">test_folds</span><span class="p">),</span> <span class="n">dists</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">random_simple_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
      <span class="n">test_size</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Distances train/test: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">train/train: </span><span class="si">%f</span><span class="se">\t</span><span class="s2">test/test: </span><span class="si">%f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">dists</span><span class="p">)</span>

<span class="n">train_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">train_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">test_folds</span><span class="p">,</span> <span class="n">subset_name</span><span class="o">=</span><span class="s2">&quot;Train&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Training set&quot;</span><span class="p">)</span>
<span class="n">train_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Testing set&quot;</span><span class="p">)</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
</pre></div>
</div>
<p>Let’s fit the model with a 2-fold cross-validation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Model parameters</span>
<span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;penalty&quot;</span><span class="p">:</span><span class="s1">&#39;elasticnet&#39;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">:</span><span class="mf">1.0</span><span class="p">,</span> <span class="s2">&quot;fit_intercept&quot;</span><span class="p">:</span><span class="kc">True</span><span class="p">,</span> <span class="s2">&quot;class_weight&quot;</span><span class="p">:</span><span class="s2">&quot;balanced&quot;</span><span class="p">,</span>
          <span class="s2">&quot;intercept_scaling&quot;</span><span class="p">:</span><span class="mf">1.</span><span class="p">,</span> <span class="s2">&quot;random_state&quot;</span><span class="p">:</span><span class="mi">1234</span><span class="p">,</span> <span class="s2">&quot;max_iter&quot;</span><span class="p">:</span><span class="mi">2000</span><span class="p">,</span> <span class="s2">&quot;tol&quot;</span><span class="p">:</span> <span class="mf">1e-4</span><span class="p">,</span>
          <span class="s2">&quot;multi_class&quot;</span><span class="p">:</span><span class="s1">&#39;multinomial&#39;</span><span class="p">,</span> <span class="s2">&quot;n_jobs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;l1_ratio&quot;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;solver&quot;</span><span class="p">:</span> <span class="s2">&quot;saga&quot;</span><span class="p">,</span>
          <span class="c1">## parameter subset allows to only consider Top-N features in terms of cross-sample variance for speed-up</span>
          <span class="s2">&quot;preprocessing_str&quot;</span><span class="p">:</span> <span class="s2">&quot;same_feature_preprocessing&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">}</span>

<span class="n">start</span><span class="o">=</span><span class="n">time</span><span class="p">()</span>

<span class="n">results</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">training_testing</span><span class="o">.</span><span class="n">cv_training</span><span class="p">(</span><span class="n">stanscofi</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">LogisticRegression</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span>
    <span class="n">test_dataset</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;AUC&quot;</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">nsplits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1234</span><span class="p">,</span>
    <span class="n">cv_type</span><span class="o">=</span><span class="s2">&quot;random&quot;</span><span class="p">,</span> <span class="n">show_plots</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s2">&quot;models&quot;</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">])]</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Elapsed time </span><span class="si">%.3f</span><span class="s2"> sec.&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;AUC on Test </span><span class="si">%f</span><span class="s2"> (Train </span><span class="si">%f</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="p">(</span>
    <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">]),</span> <span class="n">results</span><span class="p">[</span><span class="s2">&quot;train_metric&quot;</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;test_metric&quot;</span><span class="p">])]</span>
<span class="p">))</span>
</pre></div>
</div>
<p>Then, let’s predict the labels on the test dataset:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Predict the model on the testing dataset</span>
<span class="n">scores</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_proba</span><span class="p">(</span><span class="n">test_dataset</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">print_scores</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">print_classification</span><span class="p">(</span><span class="n">predictions</span><span class="p">)</span>
</pre></div>
</div>
<p>Let’s finally plot the validation figures:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Validate the model on the testing dataset</span>
<span class="n">metrics</span><span class="p">,</span> <span class="n">plot_args</span> <span class="o">=</span> <span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">compute_metrics</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">test_dataset</span><span class="p">,</span>
       <span class="n">metrics</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;AUC&quot;</span><span class="p">,</span> <span class="s2">&quot;Fscore&quot;</span><span class="p">),</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">plot_metrics</span><span class="p">(</span><span class="o">**</span><span class="n">plot_args</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span>
       <span class="n">model_name</span><span class="o">=</span><span class="n">model</span><span class="o">.</span><span class="n">name</span><span class="o">+</span><span class="s2">&quot; &quot;</span><span class="o">+</span><span class="n">dataset</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="n">metrics</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/LR_validation.png"><img alt="Validation plots for logistic regression" src="_images/LR_validation.png" style="width: 700px;" /></a>
<p>and check errors in predictions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Plot the model on the testing dataset</span>
<span class="n">test_dataset</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">predictions</span><span class="o">=</span><span class="n">predictions</span><span class="p">,</span> <span class="n">show_errors</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/LR_errors.png"><img alt="Visualization of the errors in predictions by logistic regression" src="_images/LR_errors.png" style="width: 700px;" /></a>
</section>
</section>
<section id="ranking-metrics">
<h2>Ranking metrics<a class="headerlink" href="#ranking-metrics" title="Permalink to this heading"></a></h2>
<p>Since version 2.0.0 of <em>stanscofi</em>, several ranking metrics have been integrated, in addition to the common accuracy and precision measures (<code class="docutils literal notranslate"><span class="pre">AUC</span></code>, <code class="docutils literal notranslate"><span class="pre">Fscore</span></code> (<span class="math notranslate nohighlight">\(F_{\beta}\)</span>), <code class="docutils literal notranslate"><span class="pre">F1K</span></code> (<span class="math notranslate nohighlight">\(F_{\beta}\)</span> at a fixed rank of prediction scores)). The corresponding list of metric names are accessible as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">stanscofi.validation</span><span class="w"> </span><span class="kn">import</span> <span class="n">metrics_list</span>
</pre></div>
</div>
<p><strong>Notation:</strong> Let us consider the ranking on <span class="math notranslate nohighlight">\(N\)</span> pairs by <em>decreasing</em> predicted score identified by <span class="math notranslate nohighlight">\(\{1,2,\dots,N\}\)</span> given by permutation <span class="math notranslate nohighlight">\(\sigma : \{p_1,p_2,\dots,p_k,p_{k+1},\dots,p_N\} \rightarrow [N]=\{1,2,\dots,N\}\)</span>, and the list of true scores <span class="math notranslate nohighlight">\(t : \{p_1,p_2,\dots,p_k\} \rightarrow \mathbb{R}\)</span>. <span class="math notranslate nohighlight">\(N_\text{pos}\)</span> is the number of positive pairs among the <span class="math notranslate nohighlight">\(N\)</span> ones.</p>
<p>In code, <span class="math notranslate nohighlight">\(y_\text{test}\)</span> is the vector of true labels (corresponding to <span class="math notranslate nohighlight">\(t\)</span>), whereas <span class="math notranslate nohighlight">\(\text{scores}\)</span> is the vector of corresponding predicted scores (corresponding to the ranking <span class="math notranslate nohighlight">\(\sigma\)</span>).</p>
<section id="dcg-ncdg">
<h3>DCG / NCDG<a class="headerlink" href="#dcg-ncdg" title="Permalink to this heading"></a></h3>
<p>Formula for normalized discounted cumulative gain (NDCG) up to rank <span class="math notranslate nohighlight">\(k&lt;N=\)</span>#pairs on the ranking on pairs by <em>decreasing</em> predicted score given by the permutation <span class="math notranslate nohighlight">\(\sigma : \{p_1,p_2,\dots,p_k,p_{k+1},\dots,p_N\} \rightarrow [N]=\{1,2,\dots,N\}\)</span>, and the list of true scores <span class="math notranslate nohighlight">\(t : \{p_1,p_2,\dots,p_k\} \rightarrow \mathbb{R}\)</span> corresponding to the ideal permutation <span class="math notranslate nohighlight">\(\sigma^\star\)</span> (such that <span class="math notranslate nohighlight">\((\sigma^\star)^{-1}(1) = \arg\max_{i}t(p_i)\)</span>, <span class="math notranslate nohighlight">\(\dots\)</span>, <span class="math notranslate nohighlight">\((\sigma^\star)^{-1}(N)=\arg\min_{i}t(p_i)\)</span>):</p>
<div class="math notranslate nohighlight">
\[\text{nDCG}_k(\sigma,t) := \underbrace{\sum_{i=1}^k \frac{t[\sigma^{-1}(i)]}{\log_2(i+1)}}_\text{DCG$_k(\sigma,t)$} / \underbrace{\sum_{j=1}^k \frac{t[(\sigma^\star)^{-1}(j)]}{\log_2(j+1)}}_\text{ideal DCG$_k(\sigma,t)$=DCG$_k(\sigma^\star,t)$}\;.\]</div>
<p>In code (at rank <code class="docutils literal notranslate"><span class="pre">k</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">DCGk</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">NDCGk</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="mean-reciprocal-rank-mrr">
<h3>Mean reciprocal rank (MRR)<a class="headerlink" href="#mean-reciprocal-rank-mrr" title="Permalink to this heading"></a></h3>
<p>Here, we only perform one ‘’query’’ (retrieve a positive pair) and as such, the corresponding rank of the correct answer in our case is the minimum rank of a positive pair in the ranking by predicted scores:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\text{MRR}(\sigma,t) := \arg\min_{\substack{r \in [N]\\t[\sigma^{-1}(r)]=1}} 1/r\;.\end{split}\]</div>
<p>In code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">MRR</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="r-precision">
<h3>R-Precision<a class="headerlink" href="#r-precision" title="Permalink to this heading"></a></h3>
<div class="math notranslate nohighlight">
\[\text{RP}(\sigma,t) := \frac{\sum_{r=1}^{N_\text{pos}} \delta(t[\sigma^{-1}(r)]=1)}{N_\text{pos}}\;.\]</div>
<p>In code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">Rscore</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="precision-k">
<h3>Precision &#64; k<a class="headerlink" href="#precision-k" title="Permalink to this heading"></a></h3>
<div class="math notranslate nohighlight">
\[\text{Prec&#64;}k(\sigma,t) := \frac{\sum_{r=1}^{k} \delta(t[\sigma^{-1}(r)]=1)}{k}\;.\]</div>
<p>In code (at rank <code class="docutils literal notranslate"><span class="pre">k</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">PrecisionK</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="average-precision-ap">
<h3>Average Precision (AP)<a class="headerlink" href="#average-precision-ap" title="Permalink to this heading"></a></h3>
<div class="math notranslate nohighlight">
\[\text{AP}(\sigma,t) := \frac{1}{N_\text{pos}}\sum_{r=1}^{N_\text{pos}}\text{Prec&#64;}r(\sigma,t)\;.\]</div>
<p>In code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">AP</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="mean-average-precision-map">
<h3>Mean Average Precision (mAP)<a class="headerlink" href="#mean-average-precision-map" title="Permalink to this heading"></a></h3>
<p>Since we only have one class, in our case mAP=AP.</p>
<p>In code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">MAP</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="kendall-s-tau">
<h3>Kendall’s <span class="math notranslate nohighlight">\(\tau\)</span><a class="headerlink" href="#kendall-s-tau" title="Permalink to this heading"></a></h3>
<p>If <span class="math notranslate nohighlight">\(N_\text{dis}\)</span> is the number of discordant pairs between predicted scores <span class="math notranslate nohighlight">\(p\)</span> and true labels <span class="math notranslate nohighlight">\(t\)</span> (i.e., <span class="math notranslate nohighlight">\((i,j)\)</span> such that <span class="math notranslate nohighlight">\(p[i]&lt;p[j] \land t[i]&gt;t[j]\)</span> or <span class="math notranslate nohighlight">\(p[j]&lt;p[i] \land t[j]&gt;t[i]\)</span>)</p>
<div class="math notranslate nohighlight">
\[\tau(p,t) = 1 - \frac{2N_\text{dis}(p,t)}{C^n_2}\;.\]</div>
<p>A smarter version is implemented in <code class="docutils literal notranslate"><span class="pre">scipy</span></code> which takes into account ties.</p>
<p>In code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">TAU</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="spearman-s-r">
<h3>Spearman’s <span class="math notranslate nohighlight">\(r\)</span><a class="headerlink" href="#spearman-s-r" title="Permalink to this heading"></a></h3>
<p>Roughly Pearson correlation on ranks instead of scores.</p>
</section>
<section id="expected-reciprocal-rank-err">
<h3>Expected reciprocal rank (ERR)<a class="headerlink" href="#expected-reciprocal-rank-err" title="Permalink to this heading"></a></h3>
<p>Here, we consider a single query (find one positive pair). If <span class="math notranslate nohighlight">\(\sigma\)</span> is the ranking induced by decreasing predicted scores and <span class="math notranslate nohighlight">\(t\)</span> the true binary labels on <span class="math notranslate nohighlight">\(N\)</span> pairs. This is supposed to introduce some dependency with prior elements encountered in the ranking (see <a class="reference external" href="https://dl.acm.org/doi/abs/10.1145/1645953.1646033">reference</a>).</p>
<div class="math notranslate nohighlight">
\[\text{ERR} := \sum_{k=1}^N \frac{1}{k}t[\sigma^{-1}(k)]\Pi_{i=1}^{k-1}(1-t[\sigma^{-1}(i)])\;.\]</div>
<p>In code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stanscofi</span><span class="o">.</span><span class="n">validation</span><span class="o">.</span><span class="n">ERR</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">scores</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="stanscofi_install.html" class="btn btn-neutral float-left" title="Installation of stanscofi" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="stanscofi_modules.html" class="btn btn-neutral float-right" title="Documentation in stanscofi" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Clémence Réda.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>